spacy/pipeline/trainable_pipe.pyx

```diff
@@ -0,0 +1,319 @@
+# cython: infer_types=True, profile=True
+from typing import Iterable, Iterator, Optional, Dict, Tuple, Callable
+import srsly
+from thinc.api import set_dropout_rate, Model, Optimizer
+
+from ..tokens.doc cimport Doc
+
+from ..training import validate_examples
+from ..errors import Errors
+from .pipe import Pipe, deserialize_config
+from .. import util
+from ..vocab import Vocab
+from ..language import Language
+from ..training import Example
+
+cdef class TrainablePipe(Pipe):
```

Comment:
I can't remember whether this needed to be a `cdef class`. Is it because we need its subclasses to be `cdef class`, and they can't inherit from Python types? I think that's it, but I just wanted to check.

Reply:
Yeah, I tested this during the refactor I think (and maybe @svlandeg did as well?) and I'm pretty sure we couldn't have the other `cdef` classes inherit from it otherwise.